libkbfs/prefetcher.go

```diff
@@ -204,46 +486,109 @@ func (p *blockPrefetcher) prefetchDirectDirBlock(ptr BlockPointer, b *DirBlock,
-func (p *blockPrefetcher) PrefetchBlock(block Block, ptr BlockPointer,
-	kmd KeyMetadata, priority int) (<-chan struct{}, <-chan struct{}, error) {
-	// TODO: Remove this log line.
-	p.log.CDebugf(context.TODO(), "Prefetching block by request from "+
-		"upstream component. Priority: %d", priority)
-	doneCh := make(chan struct{}, 1)
-	errCh := make(chan struct{}, 1)
-	return doneCh, errCh,
-		p.request(priority, kmd, ptr, block, "", doneCh, errCh)
+	if totalNumBlocks == 0 {
+		isTail = true
+	}
+	return numBlocks, isTail
 }
 
-// PrefetchAfterBlockRetrieved implements the Prefetcher interface for
-// blockPrefetcher. Returns a channel that is closed once all the prefetches
+// handlePrefetch allows the prefetcher to trigger prefetches. Whichever
+// component is responsible for retrieving blocks will call this method once
+// it's done retrieving a block. Returns `numBlocks` which indicates how many
+// blocks with a parent of `ptr.ID` must complete for `ptr.ID`'s prefetch to be
 // complete.
-func (p *blockPrefetcher) PrefetchAfterBlockRetrieved(b Block,
-	ptr BlockPointer, kmd KeyMetadata) (doneCh, errCh <-chan struct{},
-	numBlocks int) {
-	switch b := b.(type) {
+func (p *blockPrefetcher) handlePrefetch(ctx context.Context, pre *prefetch) (
+	numBlocks int, isTail bool, err error) {
+	req := pre.req
+	switch b := req.block.(type) {
 	case *FileBlock:
 		if b.IsInd {
-			doneCh, errCh, numBlocks = p.prefetchIndirectFileBlock(b, kmd)
+			numBlocks, isTail = p.prefetchIndirectFileBlock(ctx,
+				req.ptr.ID, b, req.kmd, req.lifetime)
+		} else {
+			isTail = true
 		}
 	case *DirBlock:
 		if b.IsInd {
-			doneCh, errCh, numBlocks = p.prefetchIndirectDirBlock(b, kmd)
+			numBlocks, isTail = p.prefetchIndirectDirBlock(ctx, req.ptr.ID,
+				b, req.kmd, req.lifetime)
 		} else {
-			doneCh, errCh, numBlocks = p.prefetchDirectDirBlock(ptr, b, kmd)
+			numBlocks, isTail = p.prefetchDirectDirBlock(ctx, req.ptr.ID,
+				b, req.kmd, req.lifetime)
 		}
 	default:
 		// Skipping prefetch for block of unknown type (likely CommonBlock)
+		return 0, false, errors.New("Unknown block type")
+	}
+	return numBlocks, isTail, nil
+}
+
+func (p *blockPrefetcher) triggerPrefetch(req *prefetchRequest) {
+	select {
+	case p.prefetchRequestCh.In() <- req:
+	case <-p.shutdownCh:
+		p.log.Warning("Skipping prefetch for block %v since "+
+			"the prefetcher is shutdown", req.ptr.ID)
+	}
+	return
+}
+
+func (p *blockPrefetcher) cacheOrCancelPrefetch(ctx context.Context,
+	ptr BlockPointer, tlfID tlf.ID, block Block, lifetime BlockCacheLifetime,
+	prefetchStatus PrefetchStatus) error {
+	err := p.retriever.PutInCaches(ctx, ptr, tlfID, block, lifetime,
+		prefetchStatus)
+	if err != nil {
+		p.log.CWarningf(ctx, "error prefetching block %s, canceling", ptr.ID)
+		p.CancelPrefetch(ptr.ID)
+	}
+	return err
+}
+
+// TriggerPrefetch triggers a prefetch if appropriate.
+func (p *blockPrefetcher) TriggerPrefetch(ctx context.Context,
```

Comment:
`ProcessBlockForPrefetch`?

Reply:
Sure! I think you can shorten it to just `ProcessBlock()` if you want, since it's in the `Prefetcher` interface. But your call.  (I guess it might be better to be explicit, in case some class implements it along with other interfaces in the future.)